azure/services/virtualnetworks/virtualnetworks.go

```diff
@@ -19,169 +19,65 @@ package virtualnetworks
+	svc.ConditionType = infrav1.VNetReadyCondition
+	svc.PostCreateOrUpdateResourceHook = postCreateOrUpdateResourceHook
+	return svc
 }
 
-// Delete deletes the virtual network if it is managed by capz.
-func (s *Service) Delete(ctx context.Context) error {
-	ctx, log, done := tele.StartSpanWithLogger(ctx, "virtualnetworks.Service.Delete")
-	defer done()
-
-	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
-	defer cancel()
-
-	vnetSpec := s.Scope.VNetSpec()
-	if vnetSpec == nil {
-		return nil
-	}
-
-	// Check that the vnet is not BYO.
-	managed, err := s.IsManaged(ctx)
+func postCreateOrUpdateResourceHook(ctx context.Context, scope VNetScope, existingVnet *asonetworkv1.VirtualNetwork, err error) error {
 	if err != nil {
-		if azure.ResourceNotFound(err) {
-			// already deleted or doesn't exist, cleanup status and return.
-			s.Scope.DeleteLongRunningOperationState(vnetSpec.ResourceName(), serviceName, infrav1.DeleteFuture)
-			s.Scope.UpdateDeleteStatus(infrav1.VNetReadyCondition, serviceName, nil)
-			return nil
-		}
-		return errors.Wrap(err, "could not get VNet management state")
+		return err
 	}
-	if !managed {
-		log.Info("Skipping VNet deletion in custom vnet mode")
-		return nil
-	}
-
-	err = s.DeleteResource(ctx, vnetSpec, serviceName)
-	s.Scope.UpdateDeleteStatus(infrav1.VNetReadyCondition, serviceName, err)
-	return err
-}
-
-// IsManaged returns true if the virtual network has an owned tag with the cluster name as value,
-// meaning that the vnet's lifecycle is managed.
-func (s *Service) IsManaged(ctx context.Context) (bool, error) {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "virtualnetworks.Service.IsManaged")
-	defer done()
 
-	spec := s.Scope.VNetSpec()
-	if spec == nil {
-		return false, errors.New("cannot get vnet to check if it is managed: spec is nil")
-	}
-
-	scope := azure.VNetID(s.Scope.SubscriptionID(), spec.ResourceGroupName(), spec.ResourceName())
-	result, err := s.TagsGetter.GetAtScope(ctx, scope)
+	vnet := scope.Vnet()
+	vnet.ID = ptr.Deref(existingVnet.Status.Id, "")
+	vnet.Tags = existingVnet.Status.Tags
+
+	// Update the subnet CIDRs if they already exist.
+	// This makes sure the subnet CIDRs are up to date and there are no validation errors when updating the VNet.
+	// Subnets that are not part of this cluster spec are silently ignored.
+	subnets := &asonetworkv1.VirtualNetworksSubnetList{}
+	err = scope.GetClient().List(ctx, subnets,
+		client.InNamespace(existingVnet.Namespace),
+		client.MatchingLabels{labels.OwnerNameLabel: existingVnet.Name},
+	)
 	if err != nil {
-		return false, err
+		return errors.Wrap(err, "failed to list subnets")
 	}
-
-	tagsMap := make(map[string]*string)
-	if result.Properties != nil && result.Properties.Tags != nil {
-		tagsMap = result.Properties.Tags
+	for _, subnet := range subnets.Items {
+		scope.UpdateSubnetCIDRs(subnet.AzureName(), converters.GetSubnetAddresses(subnet))
```

Comment:
I saw we also call `scope.UpdateSubnetCIDRs` in `subnets.go`. Is there a reason for doing this in two separate places?

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2339